test/addressBook.test.ts

```diff
@@ -0,0 +1,304 @@
+import {
+  MockERC20Instance,
+  WhitelistInstance,
+  OtokenFactoryInstance,
+  UpgradeableContractV1Instance,
+  UpgradeableContractV2Instance,
+  OwnedUpgradeabilityProxyInstance,
+  AddressBookInstance,
+} from '../build/types/truffle-types'
+
+const {expectRevert} = require('@openzeppelin/test-helpers')
+
+const MockERC20 = artifacts.require('MockERC20.sol')
+const OtokenFactory = artifacts.require('OtokenFactory.sol')
+const MarginPool = artifacts.require('MarginPool.sol')
+const UpgradeableContractV1 = artifacts.require('UpgradeableContractV1.sol')
+const UpgradeableContractV2 = artifacts.require('UpgradeableContractV2.sol')
+const OwnedUpgradeabilityProxy = artifacts.require('OwnedUpgradeabilityProxy.sol')
+const AddressBook = artifacts.require('AddressBook.sol')
+const Whitelist = artifacts.require('Whitelist.sol')
+
+contract('AddressBook', ([owner, otokenImplAdd, marginPoolAdd, random]) => {
+  // ERC20 mocks
+  let weth: MockERC20Instance
+  // addressbook instance
+  let addressBook: AddressBookInstance
+
+  before('Deployment', async () => {
+    // deploy WETH token
+    weth = await MockERC20.new('WETH', 'WETH')
+    // deploy AddressBook token
+    addressBook = await AddressBook.new()
+  })
+
+  describe('Set otoken implementation address', () => {
+    it('should revert adding otoken implementation address from non-owner address', async () => {
+      await expectRevert(addressBook.setOtokenImpl(otokenImplAdd, {from: random}), 'Ownable: caller is not the owner')
+    })
+
+    it('should set otoken implementation address', async () => {
+      await addressBook.setOtokenImpl(otokenImplAdd, {from: owner})
+
+      assert.equal(await addressBook.getOtokenImpl(), otokenImplAdd, 'Otoken implementation address mismatch')
+    })
+  })
+
+  describe('Set controller', () => {
+    let controller: UpgradeableContractV1Instance
+
+    before(async () => {
+      controller = await UpgradeableContractV1.new()
+    })
+
+    it('should revert adding otoken implementation address from non-owner address', async () => {
+      await expectRevert(
+        addressBook.setController(controller.address, {from: random}),
+        'Ownable: caller is not the owner',
+      )
+    })
+
+    it('should set controller address', async () => {
+      await addressBook.setController(controller.address, {from: owner})
+
+      const proxy: OwnedUpgradeabilityProxyInstance = await OwnedUpgradeabilityProxy.at(
+        await addressBook.getController(),
+      )
+      const implementationAddress = await proxy.implementation()
+
+      assert.equal(controller.address, implementationAddress, 'Controller implementation address mismatch')
+
+      controller = await UpgradeableContractV1.at(await addressBook.getController())
```

Comment:
```suggestion
```
I don't think you need this line?
controller is already `UpgradeableContractV1`, and getController() return `controller.address`

Reply:
yeah but `await UpgradeableContractV1.new()` and `await addressBook.getController()` are two different addresses, the `getController()` return the proxy address.

Reply:
Okay got it now ;)